File: sdks/node/tests/contract.test.ts
import { describe, it, expect } from "vitest";
import { readFileSync } from "node:fs";
import { resolve, dirname } from "node:path";
import { fileURLToPath } from "node:url";
import { parse as parseYaml } from "yaml";

const __dirname = dirname(fileURLToPath(import.meta.url));
const REPO_ROOT = resolve(__dirname, "..", "..", "..");
const OPENAPI_SPEC_PATH = resolve(REPO_ROOT, "openapi", "sync.openapi.yaml");
const SDK_MANIFEST_PATH = resolve(__dirname, "..", "sdk-manifest.yaml");

function loadYaml(path: string): Record<string, unknown> {
  return parseYaml(readFileSync(path, "utf-8")) as Record<string, unknown>;
}

function getSpecOperationIds(spec: Record<string, unknown>): Set<string> {
  const ids = new Set<string>();
  const paths = spec.paths as Record<string, Record<string, Record<string, unknown>>> | undefined;
  if (!paths) return ids;
  for (const methods of Object.values(paths)) {
    for (const operation of Object.values(methods)) {
      if (operation && typeof operation === "object" && "operationId" in operation) {
        ids.add(operation.operationId as string);
      }
    }
  }
  return ids;
}

describe("Contract: SDK manifest matches OpenAPI spec", () => {
  const spec = loadYaml(OPENAPI_SPEC_PATH);
  const manifest = loadYaml(SDK_MANIFEST_PATH);
  const specOperationIds = getSpecOperationIds(spec);

  it("all manifest operations exist in the OpenAPI spec", () => {
    const operations = (manifest as Record<string, unknown>).operations as Array<{
      operationId: string;
    }>;
    for (const op of operations) {
      expect(specOperationIds.has(op.operationId)).toBe(true);
    }
  });

  it("api_version matches spec version", () => {
    const sdk = (manifest as Record<string, unknown>).sdk as Record<string, unknown>;
    const specInfo = spec.info as Record<string, unknown>;
    expect(sdk.api_version).toBe(specInfo.version);
  });

  it("ProcessingResponse schema has required fields", () => {
    const schemas = (
      spec.components as Record<string, unknown>
    ).schemas as Record<string, Record<string, unknown>>;
    const responseSchema = schemas.ProcessingResponse;
    const required = responseSchema.required as string[];
    expect(required).toContain("StateText");
    expect(required).toContain("TraceId");
  });

  it("Extraction schema has expected properties", () => {
    const schemas = (
      spec.components as Record<string, unknown>
    ).schemas as Record<string, Record<string, unknown>>;
    const extractionSchema = schemas.Extraction;
    const props = Object.keys(
      extractionSchema.properties as Record<string, unknown>
    );
    expect(props).toContain("documentType");
    expect(props).toContain("confidence");
    expect(props).toContain("countryCode");
    expect(props).toContain("pageRange");
    expect(props).toContain("extraction");
  });
});

File: sdks/rust/tests/contract_test.rs
use std::collections::HashMap;
use std::fs;
use std::path::PathBuf;

fn repo_root() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("..")
        .join("..")
}

fn load_yaml(path: &std::path::Path) -> serde_yaml::Value {
    let content = fs::read_to_string(path).expect(&format!("failed to read {:?}", path));
    serde_yaml::from_str(&content).expect("failed to parse YAML")
}

#[test]
fn test_manifest_operations_exist_in_spec() {
    let root = repo_root();
    let spec = load_yaml(&root.join("openapi").join("sync.openapi.yaml"));
    let manifest = load_yaml(&PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("sdk-manifest.yaml"));

    // Collect all operationIds from spec
    let mut spec_ops: Vec<String> = Vec::new();
    if let Some(paths) = spec.get("paths").and_then(|v| v.as_mapping()) {
        for (_path, methods) in paths {
            if let Some(method_map) = methods.as_mapping() {
                for (_method, operation) in method_map {
                    if let Some(op_id) = operation.get("operationId").and_then(|v| v.as_str()) {
                        spec_ops.push(op_id.to_string());
                    }
                }
            }
        }
    }

    // Verify manifest operations exist in spec
    if let Some(operations) = manifest.get("operations").and_then(|v| v.as_sequence()) {
        for op in operations {
            let op_id = op.get("operationId").and_then(|v| v.as_str()).unwrap();
            assert!(
                spec_ops.contains(&op_id.to_string()),
                "manifest operationId '{}' not found in OpenAPI spec. Available: {:?}",
                op_id,
                spec_ops
            );
        }
    }
}

#[test]
fn test_api_version_matches() {
    let root = repo_root();
    let spec = load_yaml(&root.join("openapi").join("sync.openapi.yaml"));
    let manifest = load_yaml(&PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("sdk-manifest.yaml"));

    let spec_version = spec["info"]["version"].as_str().unwrap();
    let manifest_version = manifest["sdk"]["api_version"].as_str().unwrap();
    assert_eq!(spec_version, manifest_version);
}

#[test]
fn test_processing_response_schema_has_required_fields() {
    let root = repo_root();
    let spec = load_yaml(&root.join("openapi").join("sync.openapi.yaml"));

    let required = spec["components"]["schemas"]["ProcessingResponse"]["required"]
        .as_sequence()
        .unwrap();
    let required_strs: Vec<&str> = required.iter().filter_map(|v| v.as_str()).collect();

    assert!(required_strs.contains(&"StateText"));
    assert!(required_strs.contains(&"TraceId"));
}

#[test]
fn test_extraction_schema_has_expected_properties() {
    let root = repo_root();
    let spec = load_yaml(&root.join("openapi").join("sync.openapi.yaml"));

    let properties = spec["components"]["schemas"]["Extraction"]["properties"]
        .as_mapping()
        .unwrap();

    let props: Vec<&str> = properties
        .keys()
        .filter_map(|k| k.as_str())
        .collect();

    assert!(props.contains(&"documentType"));
    assert!(props.contains(&"confidence"));
    assert!(props.contains(&"countryCode"));
    assert!(props.contains(&"pageRange"));
    assert!(props.contains(&"extraction"));
}
